HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc

```diff
@@ -0,0 +1,390 @@
+      FailedL2 = true;
+
+    //pt difference cut
+    double L2candPt = L2cand->pt();
+    double L2trackPt = L2tk->pt();
+
+    if (std::abs(L2candPt - L2trackPt) > max_PtDifference_)
+      FailedL2 = true;
+
+    //track pt cut
+    if (L2trackPt < min_TrackPt_)
+      FailedL2 = true;
+
+    // Pt threshold cut
+    double L2pt = L2cand->pt();
+    double L2err0 = L2tk->error(0);
+    double L2abspar0 = std::abs(L2tk->parameter(0));
+    double L2ptLx = L2pt;
+    // convert 50% efficiency threshold to 90% efficiency threshold
+    if (L2abspar0 > 0)
+      L2ptLx += nsigma_Pt_ * L2err0 / L2abspar0 * L2pt;
+    LogTrace("HLTMuonL3orL2PreFilter") << " ...Muon in loop, trackkRef pt= " << L2tk->pt() << ", ptLx= " << L2ptLx
+                                       << " cand pT " << L2cand->pt();
+    if (L2ptLx < min_Pt_L2_)
+      FailedL2 = true;
+
+    bool matchedL3 = false;
+
+    ////////////////////////////////////////L3///////////////////////////////////////////////////////////
+    for (unsigned int iMu = 0; iMu < mucands.size(); iMu++) {
+      reco::RecoChargedCandidateRef cand(iEvent.getHandle(candToken_), iMu);
+      LogDebug("HLTMuonL3SimplePreFilter") << "cand isNonnull " << cand.isNonnull();
+
+      //did this candidate triggered at previous stage.
+      if (matchPreviousCand_ && !triggeredByPreviousLevel(cand, vcands))
+        continue;
+
+      if (std::abs(cand->eta()) > max_Eta_L3_)
+        continue;
+
+      reco::TrackRef tk = cand->track();
+      LogDebug("HLTMuonL3SimplePreFilter")
+          << " Muon in loop, q*pt= " << tk->charge() * tk->pt() << " (" << cand->charge() * cand->pt() << ") "
+          << ", eta= " << tk->eta() << " (" << cand->eta() << ") "
+          << ", hits= " << tk->numberOfValidHits() << ", d0= " << tk->d0() << ", dz= " << tk->dz();
+
+      // cut on number of hits
+      if (tk->numberOfValidHits() < min_Nhits_L3_)
+        continue;
+
+      //normalizedChi2 cut
+      if (tk->normalizedChi2() > max_NormalizedChi2_L3_)
+        continue;
+
+      if (not recoBeamSpotHandle.isValid()) {
+        throw cms::Exception("MissingProduct") << "can not find " << beamspotTag_;
+      }  //Throw an exception if the beamSpot is not valid
+
+      if (recoBeamSpotHandle.isValid()) {
+        const reco::BeamSpot& beamSpot = *recoBeamSpotHandle;
+
+        //dz cut (impact parameter of the muon in the z direction computed with respect to the beamspot position)
+        float vz_L3 = cand->vz() - beamSpot.z0();
+        float vx_L3 = cand->vx() - beamSpot.x0();
+        float vy_L3 = cand->vy() - beamSpot.y0();
+        if (std::abs(vz_L3 - (vx_L3 * cand->px() + vy_L3 * cand->py()) / (cand->pt() * cand->pt()) * cand->pz()) >
+            max_Dz_)
+          continue;
+
+        // dxy significance cut (safeguard against bizarre values)
+        if (min_DxySig_ > 0 &&
+            (tk->dxyError() <= 0 || std::abs(tk->dxy(beamSpot.position()) / tk->dxyError()) < min_DxySig_))
```

Comment:
I don't fully understand this change. In the original one, we are comparing the significance of the track with the minimum significance specified in the parameter set.

Reply:
The idea was just to have `A < B*C` instead of `A/B < C`, to avoid the edge case where `B` is zero and causes a crash in the 2nd implementation (there is no check that B is >0). My assumption is that "B" (i.e. `tk->dxyError()`) is always non-negative; if not, this suggestion would not apply anymore.